File: ccore/c3dclasses/ccore/cconstructs/cresource/CResource.js
//------------------------------------------------------------------------------------
// file: cresource.js
// desc: opens input and/or output files that can be used globally throughout the SDK
//------------------------------------------------------------------------------------

//-------------------------------------------------------
// name: CResource
// desc: input and output file resource
//-------------------------------------------------------
var CResource = new Class({
	// members
	initialize : function() { 
		this.m_hashparams=null;
	}, // end CResource() 
	
	open : function( strpath, params ) {
		this.m_hashparams = new CHash();
		if( !this.m_hashparams )
			return false;
		this.m_hashparams.create( params );
		this.m_hashparams.set("cresource_path", strpath);
		return true;
	}, // end open()
	
	restore : function() { 
		return this.open( this.m_hashparams.get( "cresource_path" ), this.m_hashparams.valueOf() ); 
	}, // end restore()
	
	_toString : function() { 
		return ""; 
	}, // end toString() 
	
	getParams : function() {
		return this.m_hashparams;
	}, // end getParams()
	
	param : function(strname){
		return ( this.m_hashparams ) ? this.m_hashparams.get(strname) : ""; 
	}, // end params

	path : function(){
		return this.param( "cresource_path" ); 
	}, // end path()
	
	type : function(){
		return this.param("cresource_type"); 
	}, // end path()
	
	id : function(){
		return this.param("cresource_id"); 
	}, // end id()
	
	updateParams : function(params){
		if( params && this.m_hashparams)
			for( var key in params )
				this.m_hashparams.set( key, params[key] );
	}, // end updateParams()
	
	/////////////////////////
	// class methods
	
	ClassMethods: {
		m_hashpathtoresource : null,	// stores the "filepath" -> resource
		m_hashidtoresource : null,	// stroes the "identifier" -> resource
	
		_getByPath : function( strpath ){
			return (CResource.m_hashpathtoresource) ? CResource.m_hashpathtoresource.get( strpath ) : null;
		}, // end _getByPath()
	
		_addByPath : function( strpath, cresource ){
			if( CResource.m_hashpathtoresource == null )
				CResource.m_hashpathtoresource = new CHash();
			if( cresource && CResource.m_hashpathtoresource.containsKey( strpath ) == false ){
				CResource.m_hashpathtoresource.set( strpath, cresource );
				return true;
			} // end if
			return false;			
		}, // end _addByPath()
		
		_getByID : function( strid ){
			return (CResource.m_hashidtoresource) ? CResource.m_hashidtoresource.get( strid ) : null;
		}, // end _getByID()
	
		_addByID : function( strid, cresource ){
			if( CResource.m_hashidtoresource == null )
				CResource.m_hashidtoresource = new CHash();
			if( cresource && CResource.m_hashidtoresource.containsKey( strid ) == false ){
				CResource.m_hashidtoresource.set( strid, cresource );
				return true;
			} // end if
			return false;
		}, // end _addByID()
	
		_register : function( strid, strpath, params ){
			var cresource = null;
			if( strid && ( cresource = CResource._getByID( strid ) ) ){
				cresource.updateParams( params );
				return cresource;
			}
			if( strpath && ( cresource = CResource._getByPath( strpath ) ) ){
				CResource._addByID( strid, cresource );
				cresource.updateParams( params );
				return cresource;
			} // end if
			var strtype = params["cresource_type"];
			
			if( strtype == "" || (cresource = new window[strtype]()) == null || cresource.open( strpath, params ) == false ){
				return null; 	
			} // end if
			
			console.log("cresource: of type: " + strtype + " id: " + strid);
			console.log(cresource);
			
			CResource._addByID( strid, cresource );
			CResource._addByPath( strpath, cresource );		
			return cresource;
		}, // end _register()
	
		toStringVisit : function( strcallback ){
			var rs = CResource.m_hashidtoresource;
			return ( rs ) ? rs.toStringVisit( strcallback ) : null;
		} // end toStringVisit()
	} // end ClassMethods
}); // end CResource

function include_resource(strid, strpath, params){
	return CResource._register( strid, strpath, params );
} // end include_resource()

function use_resource(strid){
	var cresource = CResource._getByID(strid);
	return cresource;
} // end use_resource()

File: ccore/c3dclasses/ccore/cconstructs/cresource/CResource.java
//---------------------------------------------------------------------------------------
// file: CResource
// desc: an object treated as a resource or (singleton) to be used throughout the SDK  
//---------------------------------------------------------------------------------------
package c3dclasses;

//---------------------------------------------------------------------------------------
// name: CResource
// desc: an object treated as a resource or (singleton) to be used throughout the SDK  
//---------------------------------------------------------------------------------------
public class CResource extends CObject {
	public CResource() {} 
	
	// methods 
	public boolean open(String strpath, String strtype, CHash params) {
		this.append(params);
		this.set("cresource_path", strpath);
		this.set("cresource_type", strtype);
		return true;
	} // end open()
	public boolean close() { this.clear(); return true; } 
	public boolean restore() { return true; }
	public String path() { return (String) this.get("cresource_path"); }
	public String type(){ return (String) this.get("cresource_type"); }
	public String toString(){ return super.toString(); }
	
	// ClassMethods
	// include / use
	public static CResource include(String strid, String strpath, String strtype, CHash params) { 
		return CResource._register(strid, strpath, strtype, params); 
	} // end include()
	public static CResource use(String strid) { return CResource._getByID(strid); }
	// no need to call these methods below
	protected static CHash m_pathtoresource = new CHash();	// stores the "filepath" -> resource
	protected static CHash m_idtoresource = new CHash();	// stroes the "identifier" -> resource
	protected static CResource _getByPath (String strpath) { 
		return (CResource) CResource.m_pathtoresource._(strpath); 
	} // _getByPath()
	protected static boolean _addByPath (String strpath, CResource cresource) {
		if(cresource == null || CResource.m_pathtoresource.containsKey(strpath) != false)
			return false;
		CResource.m_pathtoresource._(strpath, cresource);
		return true;			
	} // end _addByPath()		
	protected static CResource _getByID (String strid) { return (CResource) CResource.m_idtoresource.get(strid); }
	protected static boolean _addByID (String strid, CResource cresource) {
		if(cresource == null && CResource.m_idtoresource.containsKey(strid) != false)
			return false;
		CResource.m_idtoresource.set(strid, cresource);
		return true;
	} // end _addByID()
	protected static CResource _register (String strid, String strpath, String strtype, CHash params) {
		CResource cresource = null;
		if(strid != null && (cresource = CResource._getByID(strid)) != null) {
			cresource.append(params);
			return cresource;
		} // end if
		if(strpath != null && (cresource = CResource._getByPath(strpath)) != null) {
			CResource._addByID(strid, cresource);
			cresource.append(params);
			return cresource;
		} // end if
		if(strtype == "" || (cresource = (CResource) __._new(strtype)) == null || 
			cresource.open(strpath, strtype, params) == false) {	
			return null; 	
		}
		CResource._addByID(strid, cresource);
		CResource._addByPath(strpath, cresource);		
		return cresource;
	} // end _register()
} // end CResource